Rename workspaces check if out name exists

DavidFair opened the pull request:
Description of work.
Modified RenameWorkspace and RenameWorkspaces to check if any of the output names are used in other workspaces
**To test:**
Renaming a workspace to the same name as an existing workspace should throw an error
Renaming workspaces should throw an error if any of the output names match an existing name

Fixes #16263 .

Release notes have been updated

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

FedeMPouzols commented on Framework/Algorithms/src/RenameWorkspaces.cpp:
```diff
@@ -86,6 +96,18 @@ void RenameWorkspaces::exec() {
   }
 
+  // Check all names are not used already before starting rename
+  for (size_t i = 0; i < nWs; ++i) {
```
I think here you could use a range-based loop: `for (const auto &name : newWsName) { ...`

FedeMPouzols commented on Framework/Algorithms/src/RenameWorkspace.cpp:
```diff
@@ -44,13 +52,28 @@ void RenameWorkspace::exec() {
       getPropertyValue("OutputWorkspace")) {
     throw std::invalid_argument(
         "The input and output workspace names must be different");
   }
```
This check could be done at the `validateInputs()` step.

FedeMPouzols commented on Framework/Algorithms/src/RenameWorkspace.cpp:
```diff
@@ -44,13 +52,28 @@ void RenameWorkspace::exec() {
+      throw std::runtime_error("The workspace " + outputwsName +
+                               " already exists");
+    }
+  }
```
And probably this too can go in `validateInputs()`?

FedeMPouzols commented:
@DavidFair: I've made some comments for possible improvements,. let me know what you think. But I think this looks good and ready to go. It doesn't break backwards compatibility, code looks good, and the documentation and examples are very good.

DavidFair commented:
I'll move the examples into validateInputs .
Previously the code also deleted the target workspace at execution time to make sure however we later discovered this caused a race condition where Poco would delete the workspace handle which we then returned so now we let the ADS handle it.

DavidFair commented:
@FedeMPouzols I've moved all tests (except one) into validate inputs and reworked the unit tests to use these now.
In RenameWorkspaces the check to see if names are in use has to be done at execution time as we need to join suffix + name + prefix then query the ADS for each one.

FedeMPouzols commented on docs/source/release/v3.8.0/framework.rst:
```diff
@@ -49,6 +49,10 @@ Improved
 
+- :ref:`RenameWorkspace <algm-RenameWorkspace>` and `RenameWorkspaces <algm-RenameWorkspaces>`
+  now check if a Workspace with that name already exists in the ADS and gives
+  the option to override it.
```
override => overwrite? If you modify this we don't need to wait for the builds again.

FedeMPouzols commented:
Looks good. It implements the plan in the issue description, code looks good, it is well covered with new/extended unit tests, and the documentation reads well.
Keeping a default of this new option as "yes, overwrite" makes perfect sense for backwards compatibility.

:shipit:

martyngigg commented:
@DavidFair This branch has developed a merge conflict
